libs/grok/src/test/java/org/elasticsearch/grok/GrokTests.java

```diff
@@ -205,6 +205,33 @@ public void testNoNamedCaptures() {
         assertEquals(expected, actual);
     }
 
+    public void testCircularReference() {
+        Exception e = expectThrows(IllegalArgumentException.class, () -> {
+            Map<String, String> bank = new HashMap<>();
+            bank.put("NAME", "!!!%{NAME}!!!");
+            String pattern = "%{NAME}";
+            new Grok(bank, pattern, false);
+        });
+        assertEquals("circular reference in pattern back [NAME][!!!%{NAME}!!!]", e.getMessage());
+
+        e = expectThrows(IllegalArgumentException.class, () -> {
+            Map<String, String> bank = new HashMap<>();
+            bank.put("NAME", "!!!%{NAME:name}!!!");
```

Comment:
if this is an attempt to catch more things... maybe add an example with type coercion as well?

```
bank.put("NAME", "!!!%{NAME:name:int}!!!");
```